software/tracksight/frontend/lib/realtime/socket.ts

```diff
@@ -0,0 +1,56 @@
+import io from "socket.io-client";
+
+import subscribeToSignal from "@/lib/api/subscribeToSignal";
+import unsubscribeFromSignal from "@/lib/api/unsubscribeFromSignal";
+import { API_BASE_URL, IS_DEBUG, MAX_RECCONECTION_ATTEMPTS } from "@/lib/constants";
+
+const socket = io(API_BASE_URL, {
+  transports: ["websocket", "polling"],
+  autoConnect: true,
+  reconnectionAttempts: MAX_RECCONECTION_ATTEMPTS,
+  timeout: 10_000,
+});
+
+export const pendingSubscriptions = new Set<string>();
+export const pendingUnsubscriptions = new Set<string>();
+
+const processPendingOperations = async () => {
+  if (pendingSubscriptions.size > 0) {
+    IS_DEBUG &&
+      console.log(
+        `%c[Socket] %cProcessing ${pendingSubscriptions.size} pending subscriptions`,
+        "color: #a3be8c; font-weight: bold;",
+        "color: #88c0d0;"
+      );
+
+    Array.from(pendingSubscriptions).forEach(subscribeToSignal);
+    pendingSubscriptions.clear();
+  }
+
+  if (pendingUnsubscriptions.size > 0) {
+    IS_DEBUG &&
+      console.log(
+        `%c[Socket] %cProcessing ${pendingUnsubscriptions.size} pending unsubscriptions`,
+        "color: #a3be8c; font-weight: bold;",
+        "color: #88c0d0;"
+      );
+
+    const unsubscriptionsToProcess =
+      Array.from(pendingUnsubscriptions).forEach(unsubscribeFromSignal);
+
+    pendingUnsubscriptions.clear();
+  }
+};
```

Comment:
the API for this is different now, so we don't need to worry about pending subs

Reply:
this is probably good to have to restore all signal connections from localstorage (when that gets implemented), but I don't think we should be calling add sub function when there is no socket connection even.

Reply:
I think this was just a lazy fix for widgets trying to subscribe to data before the socket was initialized so I'll just go back and rework it with a loading state or something for before the socket connects.

Reply:
yeah probably some init function on socket connection is sufficient for this